numpyro/handlers.py

```diff
@@ -547,7 +547,10 @@ def __init__(self, fn=None, rng_seed=None):
 
     def process_message(self, msg):
         if (msg['type'] == 'sample' and not msg['is_observed'] and
-                msg['kwargs']['rng_key'] is None) or msg['type'] == 'control_flow':
+                msg['kwargs']['rng_key'] is None) or msg['type'] in ['plate', 'control_flow']:
+            # no need to create a new key when value is available
+            if msg['value'] is not None:
+                return
```

Comment:
This saves some unnecessary splitting random keys.

Reply:
So do I understand correctly that in numpyro, in contrast to pyro, `plate` does not consume randomness? Note in Pyro it is `plate` that creates a `_Subsample` site that consumes randomness.

Reply:
@fritzo IIUC your question, then in NumPyro, `_subsample_fn` plays a similar role. If `value != None`, we will use the random key here to generate subsample indices.

Reply:
thanks for explaining!